File: src/test/java/org/jesperancinha/rockstarts/rockstarsmanager/cucumber/ArtistServiceStepdefs.java
package org.jesperancinha.rockstarts.rockstarsmanager.cucumber;

import io.cucumber.datatable.DataTable;
import io.cucumber.java.en.Given;
import io.cucumber.java.en.Then;
import io.cucumber.java.en.When;
import org.jesperancinha.rockstarts.rockstarsmanager.data.ArtistDto;
import org.jesperancinha.rockstarts.rockstarsmanager.model.Artist;
import org.jesperancinha.rockstarts.rockstarsmanager.repository.ArtistsRepository;
import org.jesperancinha.rockstarts.rockstarsmanager.services.ArtistsService;
import org.jesperancinha.rockstarts.rockstarsmanager.services.ArtistsServiceImpl;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

import static org.assertj.core.api.Assertions.assertThat;
import static org.jesperancinha.rockstarts.rockstarsmanager.converters.ArtistsConverter.toData;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

public class ArtistServiceStepdefs {

    private final ArtistsRepository artistsRepository = mock(ArtistsRepository.class);

    private final ArtistsService artistsService = new ArtistsServiceImpl(artistsRepository);

    private List<ArtistDto> artistDtos;

    private final List<ArtistDto> results = new ArrayList<>();

    @Given("^the following artist:$")
    public void the_following_artist(DataTable artistDtos) {
        this.artistDtos =artistDtos
                .asLists()
                .stream()
                .filter(artist -> !artist.get(0).equals("id")).map(artist -> new ArtistDto(Long.parseLong(artist.get(0)), artist.get(1)))
                .collect(Collectors.toList());

    }

    @When("^calling save method$")
    public void calling_save_method() {
        artistDtos.forEach(artistDto -> {
            Artist artist = toData(artistDto);
            when(artistsRepository.save(artist)).thenReturn(artist);
            ArtistDto result = artistsService.saveArtist(artistDto);
            results.add(result);
        });
    }

    @Then("^we get artist with names:$")
    public void we_get_artist_with_names(List<String> expectedArtistNames) {
        assertThat(expectedArtistNames).hasSameElementsAs(expectedArtistNames);
        for (int i = 0; i < artistDtos.size(); i++) {
            assertThat(artistDtos.get(i)
                .getName()).isEqualTo(expectedArtistNames.get(i));
        }
    }

}

File: src/test/kotlin/org/jesperancinha/rockstarts/rockstarsmanager/cucumber/ArtistServiceStepdefs.kt
package org.jesperancinha.rockstarts.rockstarsmanager.cucumber

import io.cucumber.datatable.DataTable
import io.cucumber.java.en.Given
import io.cucumber.java.en.Then
import io.cucumber.java.en.When
import org.assertj.core.api.Assertions
import org.jesperancinha.rockstarts.rockstarsmanager.converters.ArtistsConverter.toData
import org.jesperancinha.rockstarts.rockstarsmanager.data.ArtistDto
import org.jesperancinha.rockstarts.rockstarsmanager.repository.ArtistsRepository
import org.jesperancinha.rockstarts.rockstarsmanager.services.ArtistsService
import org.mockito.Mockito
import java.util.function.Consumer
import java.util.stream.Collectors

class ArtistServiceStepdefs {
    private val artistsRepository = Mockito.mock(ArtistsRepository::class.java)
    private val artistsService = ArtistsService(artistsRepository)
    private var artistDtos: List<ArtistDto>? = null
    private val results: MutableList<ArtistDto?> = ArrayList()
    @Given("^the following artist:$")
    fun the_following_artist(artistDtos: DataTable) {
        this.artistDtos = artistDtos
            .asLists()
            .stream()
            .filter { artist: List<String> -> artist[0] != "id" }
            .map { artist: List<String> -> ArtistDto(artist[0].toLong(), artist[1]) }
            .collect(Collectors.toList())
    }

    @When("^calling save method$")
    fun calling_save_method() {
        artistDtos!!.forEach(Consumer { artistDto: ArtistDto? ->
            val artist = toData(artistDto!!)
            Mockito.`when`(artistsRepository.save(artist)).thenReturn(artist)
            val result = artistsService.saveArtist(artistDto)
            results.add(result)
        })
    }

    @Then("^we get artist with names:$")
    fun we_get_artist_with_names(expectedArtistNames: List<String>) {
        Assertions.assertThat(expectedArtistNames).hasSameElementsAs(expectedArtistNames)
        for (i in artistDtos!!.indices) {
            assertThat(
                artistDtos!![i]
                    .getName()
            ).isEqualTo(expectedArtistNames[i])
        }
    }
}